### PR TITLE
roslyn-ls:  4.10.0-2.24124.2 -> 4.12.0-1.24378.2

### PR DESCRIPTION
**Pull request by konradmalik**
## Description of changes

An update and proposition to move to nuget-based distribution.

Because the fix for darwin build in the upstream repo is nowhere in sight, could we try a different approach here?

In this PR we move to installing roslyn-ls from the Nuget packages published to [this feed](https://dev.azure.com/azure-public/vside/_artifacts/feed/vs-impl) - those are official Microsoft builds.

This is the same approach that VSCode's C# Dev Kit extension uses to install this server based on: https://github.com/dotnet/roslyn/issues/71474#issuecomment-2177303207.

Not sure if this approach is accepted in nixpkgs, but if it is, then it would simplify this package quite a bit. The downside is that we would rely on binary blobs published to nuget. On the other hand, we already rely on them when downloading nuget dependencies for any dotnet project.

If this is accepted, then we can close https://github.com/NixOS/nixpkgs/pull/318497

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/ro/roslyn-ls/package.nix [hunk 1]
```diff
@@ -1,87 +1,56 @@
-{ lib, fetchFromGitHub, buildDotnetModule, dotnetCorePackages, stdenvNoCC, testers, roslyn-ls, jq }:
+{ lib
+, buildDotnetModule
+, dotnetCorePackages
+, stdenvNoCC
+, testers
+, roslyn-ls
+,
```
Comment (SuperSandro2000) [inline pkgs/by-name/ro/roslyn-ls/package.nix:7]:
```suggestion
```

File: pkgs/by-name/ro/roslyn-ls/package.nix [hunk 2]
```diff
@@ -1,87 +1,56 @@
...
-  pname = "roslyn-ls";
-  # see https://github.com/dotnet/roslyn/blob/main/eng/targets/TargetFrameworks.props
-  dotnet-sdk = with dotnetCorePackages; combinePackages [ sdk_6_0 sdk_7_0 sdk_8_0 ];
+  dotnet-sdk = dotnetCorePackages.sdk_8_0;
   # need sdk on runtime as well
-  dotnet-runtime = dotnetCorePackages.sdk_8_0;
+  dotnet-runtime = dotnet-sdk;
 
   project = "Microsoft.CodeAnalysis.LanguageServer";
+  projectLower = lib.strings.toLower project;
```
Comment (SuperSandro2000) [inline pkgs/by-name/ro/roslyn-ls/package.nix:15]:
```suggestion
  projectLower = lib.toLower project;
```

File: pkgs/by-name/ro/roslyn-ls/package.nix (hunk 1 above)
Comment (konradmalik) [reply pkgs/by-name/ro/roslyn-ls/package.nix]: b826e24611ac226a2b8bb59404bad8f60dca57fe

File: pkgs/by-name/ro/roslyn-ls/package.nix (hunk 2 above)
Comment (konradmalik) [reply pkgs/by-name/ro/roslyn-ls/package.nix]: b826e24611ac226a2b8bb59404bad8f60dca57fe

**Review by corngood — CHANGES_REQUESTED** [review]
I think it's a mistake to downgrade a package from source-built to binary, especially if there's reason to expect it to be fixed upstream.

I'll take a look at the darwin problem, but if it can't be solved, I think I'd rather see a separate `csharp-ls-bin` package added.

> On the other hand, we already rely on them when downloading nuget dependencies for any dotnet project.

I hope this will improve, but it's going to require adding add source-built dependencies to nixpkgs.

**Comment by konradmalik** [comment]
> I think it's a mistake to downgrade a package from source-built to binary, especially if there's reason to expect it to be fixed upstream.
> 
> I'll take a look at the darwin problem, but if it can't be solved, I think I'd rather see a separate `csharp-ls-bin` package added.
> 
> > On the other hand, we already rely on them when downloading nuget dependencies for any dotnet project.
> 
> I hope this will improve, but it's going to require adding add source-built dependencies to nixpkgs.

That makes sense.
I'd be grateful if you or someone more knowledgeable than me could look at the bug. It's upstream and described here: https://github.com/dotnet/roslyn/issues/73439. Maybe we could somehow solve it in nixpkgs.

I think it's not a priority for the upstream since to release packages for different architectures they most probably still use linux or windows runners but just passing different RID. Unless I'm mistaken - we can't use the same approach in nixpkgs.

**Comment by konradmalik** [comment]
Closing in favor of a fixed https://github.com/NixOS/nixpkgs/pull/318497

**Pull request closed without merging**